src/main/java/org/junitpioneer/jupiter/RangeSourceProvider.java

```diff
@@ -0,0 +1,131 @@
+		primitiveMappers.put(Long.class, Number::longValue);
+		primitiveMappers.put(Float.class, Number::floatValue);
+		primitiveMappers.put(Double.class, Number::doubleValue);
+	}
+
+	@Override
+	public Stream<? extends Arguments> provideArguments(ExtensionContext context) throws Exception {
+		// @formatter:off
+		// since it's a method annotation, the element will always be present
+		List<Annotation> argumentsSources = context
+				.getElement()
+				.map(method -> Arrays
+						.stream(method.getAnnotations())
+						.filter(annotations -> Arrays
+								.stream(annotations.annotationType().getAnnotationsByType(ArgumentsSource.class))
+								.anyMatch(annotation -> getClass().equals(annotation.value())))
+						.collect(Collectors.toList()))
+				.get();
+
+		Preconditions.condition(
+				argumentsSources.size() == 1,
+				() -> String.format(
+						"Expected exactly one annotaiton to provide an ArgumentSource, found %d.",
+						argumentsSources.size()));
+		// @formatter:on
+
+		Annotation argumentsSource = argumentsSources.get(0);
+		Class<? extends Annotation> argumentsSourceClass = argumentsSource.annotationType();
+
+		Method stepMethod = argumentsSourceClass.getMethod("step");
+		Number stepValue = (Number) stepMethod.invoke(argumentsSource);
+		double stepDouble = stepValue.doubleValue();
+		Preconditions.condition(stepDouble != 0.0, "Illegal range. The step cannot be 0.");
+
+		Method fromMethod = argumentsSourceClass.getMethod("from");
+		Number fromValue = (Number) fromMethod.invoke(argumentsSource);
+		double fromDouble = fromValue.doubleValue();
+
+		Method toMethod = argumentsSourceClass.getMethod("to");
+		Number toValue = (Number) toMethod.invoke(argumentsSource);
+		double toDouble = toValue.doubleValue();
+
+		// @formatter:off
+		Preconditions.condition(
+				fromDouble != toDouble,
+				"Illegal range. Equal from and to will produce an empty range.");
+
+		Preconditions.condition(
+				fromMethod.getReturnType() == toMethod.getReturnType(),
+				() -> String.format(
+						"Illegal range. Different classes of from (%s) and to (%s).",
+						fromMethod.getReturnType(), toMethod.getReturnType()));
+
+		Preconditions.condition(
+				fromDouble < toDouble && stepDouble > 0.0 || fromDouble > toDouble && stepDouble < 0.0,
+				() -> String.format(
+						"Illegal range. There's no way to get from %f to %f with a step of %f.",
+						fromDouble, toDouble, stepDouble));
+		// @formatter:on
+
+		Method closeMethod = argumentsSourceClass.getMethod("closed");
+		Boolean closedValue = (Boolean) closeMethod.invoke(argumentsSource);
+
+		// Once (if) Java 8 support is dropped, this boiler-plate can be cleaned up and we can just use
+		// IntStream.iterate(rangeSource.from(), i -> i < rangeSource.to(), i -> i += rangeSource.step());
+		double val = fromDouble;
+		while (val < toDouble && stepDouble > 0.0 || val > toDouble && stepDouble < 0.0
+				|| closedValue && val == toDouble) {
+			values.add(val);
```

Comment:
Instead of putting all (of possibly very many) elements into a list, I would create the stream directly. If the `Range` interface I mentioned above implements `Iterator<Number>` (that should be straight-forward) you can turn it into a stream as follows:

```java
StreamSupport.stream(
		Spliterators.spliterator(sourceIterator, size, Spliterator.ORDERED),
		false);
```

where `size` is roughly `(to - from) / step` (with some rounding).